cou/steps/analyze.py

```diff
@@ -0,0 +1,488 @@
+                    "The expected is '%s'."
+                ),
+                self.name,
+                self.os_origin,
+                "distro",
+            )
+            change_openstack_release["distro"].add(self.name)
+        elif self.os_origin not in (expected_os_origin, "distro"):
+            change_openstack_release[expected_os_origin].add(self.name)
+            logging.warning(
+                (
+                    "App: '%s' has unexpected openstack-origin or source: '%s'. "
+                    "The expected is '%s'."
+                ),
+                self.name,
+                self.os_origin,
+                expected_os_origin,
+            )
+        return change_openstack_release
+
+
+def get_openstack_release(unit: str, model_name: Union[str, None] = None) -> Union[str, None]:
+    """Return the openstack release codename based on /etc/openstack-release."""
+    cmd = "grep -Po '(?<=OPENSTACK_CODENAME=).*' /etc/openstack-release"
+    try:
+        out = model.run_on_unit(unit, cmd, model_name=model_name, timeout=20)
+    except model.CommandRunFailed:
+        logging.debug("Fall back to version check for OpenStack codename")
+        return None
+    return out["Stdout"]
+
+
+def get_pkg_version(unit: str, pkg: str, model_name: Union[str, None] = None) -> str:
+    """Get package version of a specific package in a unit."""
+    cmd = f"dpkg-query --show --showformat='${{Version}}' {pkg}"
+    out = model.run_on_unit(unit, cmd, model_name=model_name, timeout=20)
+    return out["Stdout"]
+
+
+def generate_model() -> set[Application]:
+    """Generate the applications model."""
+    juju_status = get_full_juju_status()
+    model_name = juju_status.model.name
+    apps = {
+        Application(
+            name=app,
+            status=app_status,
+            config=model.get_application_config(app),
+            model_name=model_name,
+        )
+        for app, app_status in juju_status.applications.items()
+    }
+    # NOTE(gabrielcocenza) Not all openstack-charms are mapped in the zaza lookup.
+    supported_os_charms = CHARM_TYPES.keys()
+    openstack_apps = {app for app in apps if app.charm in supported_os_charms}
+    not_supported_apps = apps - openstack_apps
+    not_supported_apps_names = sorted([app.name for app in not_supported_apps])
+    logging.warning(
+        "App(s): %s are not supported in the analyze process", ", ".join(not_supported_apps_names)
+    )
+    return openstack_apps
+
+
+def analyze() -> Analyze:
+    """Analyze the deployment before planning."""
+    logging.info("Analyzing the OpenStack deployment...")
+    apps = generate_model()
+    # E.g: {"ussuri": {"keystone"}, "victoria": {"cinder"}}
+    os_versions: defaultdict[str, set] = defaultdict(set)
+
+    # E.g {"victoria": {"keystone/0", "keystone/1"}} this means that units
+    # keystone/0 and keystone/1 should be upgraded to victoria
+    upgrade_units: defaultdict[str, set] = defaultdict(set)
```

Comment:
this set of dictionaries feels misplaced to me. If this is the analyze step, we should not be making decisions about what should happen; that should rather be in the planning phase. I think a better logical separation would be for the analyze step to only assess the status of the cloud, and for the plan step to make decisions around what should change in order to reach the goal.